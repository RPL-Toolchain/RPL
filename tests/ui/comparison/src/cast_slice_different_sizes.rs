#[cfg_attr(test, test)]
pub(crate) fn base_case() {
    let x: [i32; 3] = [1_i32, 2, 3];
    let r_x = &x;
    // Check casting through multiple bindings
    // Because it's separate, it does not check the cast back to something of the same size
    let a = r_x as *const [i32];
    let b = a as *const [u64];
    //~^ cast_slice_different_sizes
    dbg!(b);
}
#[cfg_attr(test, test)]
pub(crate) fn cross_function() {
    fn cast_slice<T, U>(slice: *const [T]) -> *const [U] {
        slice as *const [U]
        // unsafe {
        //     std::slice::from_raw_parts(
        //         slice.as_ptr() as *const U,
        //         slice.len() * std::mem::size_of::<T>() / std::mem::size_of::<U>(),
        //     )
        // }
    }
    let x: [i32; 3] = [1_i32, 2, 3];
    let r_x = &x;
    // Check casting through multiple bindings
    // Because it's separate, it does not check the cast back to something of the same size
    let b: *const [u64] = cast_slice(r_x);
    //~^ cast_slice_different_sizes
    dbg!(b);
}

pub(crate) fn main() {
    base_case();
    cross_function();
}
