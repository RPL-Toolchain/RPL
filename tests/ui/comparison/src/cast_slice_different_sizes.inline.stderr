error: casts between raw pointers to slices with different element sizes
  --> tests/ui/comparison/src/cast_slice_different_sizes.rs:8:13
   |
LL |     let a = r_x as *const [i32];
   |             ------------------- cast from this
LL |     let b = a as *const [u64];
   |             ^^^^^^^^^^^^^^^^^
   |
   = note: casting a pointer to a slice of type `i32` to a pointer to a slice of another type `u64` with different element type ABI, size or alignment can lead to undefined behavior.
   = note: `#[deny(rpl::cast_slice_different_sizes)]` on by default

error: casts between raw pointers to slices with different element sizes
  --> tests/ui/comparison/src/cast_slice_different_sizes.rs:27:27
   |
LL |     let b: *const [u64] = cast_slice(r_x);
   |                           ^^^^^^^^^^^---^
   |                                      |
   |                                      cast from this
   |
   = note: casting a pointer to a slice of type `i32` to a pointer to a slice of another type `u64` with different element type ABI, size or alignment can lead to undefined behavior.

error: aborting due to 2 previous errors

